yapapi/rest/activity.py

```diff
@@ -179,25 +178,73 @@ def id(self):
         return self._batch_id
 
 
+def _is_gsb_endpoint_not_found_error(err: ApiException) -> bool:
+    """Check if `err` is caused by "Endpoint address not found" GSB error."""
+
+    if err.status != 500:
+        return False
+    try:
+        msg = json.loads(err.body)["message"]
+        return "GSB error" in msg and "endpoint address not found" in msg
+    except Exception:
+        _log.debug("Cannot read error message from ApiException", exc_info=True)
+        return False
+
+
 class PollingBatch(Batch):
     """A `Batch` implementation that polls the server repeatedly for command status."""
 
+    async def _activity_terminated(self) -> bool:
+        """Check if the activity we're using is in "Terminated" state."""
+        try:
+            state = await self._activity.state()
+            return "Terminated" in state.state
+        except Exception:
+            _log.debug("Cannot query activity state", exc_info=True)
+            return False
+
+    async def _get_results(
+        self, timeout: float, num_tries: int, delay: float = 3.0
+    ) -> List[yaa.ExeScriptCommandResult]:
+        """Call GetExecBatchResults with re-trying on "Endpoint address not found" GSB error."""
+
+        for n in range(num_tries, 0, -1):
+            # n = num_tries, ... , 1
+            try:
+                results = await self._activity._api.get_exec_batch_results(
+                    self._activity._id, self._batch_id, _request_timeout=min(timeout, 5)
+                )
+                return results
+            except ApiException as err:
+                if await self._activity_terminated():
+                    _log.debug("Activity %s terminated by provider", self._activity._id)
```

Comment:
Changed to `warning` (as discussed with @mfranciszkiewicz).
```suggestion
                    _log.warning("Activity %s terminated by provider", self._activity._id)
```